tsdb/agent/storage_test.go

```diff
@@ -0,0 +1,325 @@
+	require.Equal(t, err, ErrUnsupported)
+}
+
+func TestExemplarQuerier(t *testing.T) {
+	opts := DefaultOptions()
+	cfg := promlog.Config{}
+	logger := promlog.New(&cfg)
+
+	s, err := NewStorage(logger, prometheus.NewRegistry(), nil, promAgentData, opts)
+	if err != nil {
+		t.Fatalf("unable to create storage for the agent: %v", err)
+	}
+
+	_, err = s.ExemplarQuerier(context.TODO())
+	require.Equal(t, err, ErrUnsupported)
+}
+
+func TestPreCommit(t *testing.T) {
+	const numDatapoints = 1000
+
+	// Create 8 series with 1000 data-points of different ranges and run queries.
+	lbls := labelsForTest()
+	opts := DefaultOptions()
+	cfg := promlog.Config{}
+	logger := promlog.New(&cfg)
+	reg := prometheus.NewRegistry()
+
+	s, err := NewStorage(logger, reg, nil, promAgentData, opts)
+	if err != nil {
+		t.Fatalf("unable to create storage for the agent: %v", err)
+	}
+
+	a := s.Appender(context.TODO())
+	for _, l := range lbls {
+		fmt.Println("label :", l)
+		lset := labels.New(l...)
+		for i := 0; i < numDatapoints; i++ {
+			sample := tsdbutil.GenerateSamples(0, 1)
+			_, err := a.Append(0, lset, sample[0].T(), sample[0].V())
+			require.NoError(t, err)
+		}
+	}
+
+	mFamily, err := reg.Gather()
+	if err != nil {
+		t.Fatalf("error gathering metrics: %v", err)
+	}
+
+	var samplesAdded float64
+	var actvieTimeSeries float64
+
+	for _, m := range mFamily {
+		if m.GetName() == "prometheus_agent_wal_active_series" {
+			actvieTimeSeries = m.Metric[0].Gauge.GetValue()
+			continue
+		}
+
+		if m.GetName() == "prometheus_agent_wal_samples_appended_total" {
+			samplesAdded = m.Metric[0].Counter.GetValue()
+			continue
+		}
+	}
+
+	require.Equal(t, samplesAdded, float64(numDatapoints*8))
+	require.Equal(t, actvieTimeSeries, float64(8))
+}
+
+func TestCommit(t *testing.T) {
+	const numDatapoints = 1000
+
+	// Create 8 series with 1000 data-points of different ranges and run queries.
+	lbls := labelsForTest()
+	opts := DefaultOptions()
+	cfg := promlog.Config{}
+	logger := promlog.New(&cfg)
+	reg := prometheus.NewRegistry()
+	remoteStorage := remote.NewStorage(log.With(logger, "component", "remote"), reg, startTime, promAgentData, time.Second*30, nil)
+
+	s, err := NewStorage(logger, reg, remoteStorage, promAgentData, opts)
+	if err != nil {
+		t.Fatalf("unable to create storage for the agent: %v", err)
+	}
+
+	a := s.Appender(context.TODO())
+	for _, l := range lbls {
+		lset := labels.New(l...)
+		for i := 0; i < numDatapoints; i++ {
+			sample := tsdbutil.GenerateSamples(0, 1)
+			_, err := a.Append(0, lset, sample[0].T(), sample[0].V())
+			require.NoError(t, err)
+		}
+	}
+
+	require.NoError(t, a.Commit())
+	time.Sleep(time.Second * 10)
+	mFamily, err := reg.Gather()
+	if err != nil {
+		t.Fatalf("error gathering metrics: %v", err)
+	}
+
+	var walPages float64
+
+	for _, m := range mFamily {
+		if m.GetName() == "prometheus_tsdb_wal_completed_pages_total" {
+			walPages = m.Metric[0].Counter.GetValue()
+			break
+		}
+	}
```

Comment:
This pattern is repeated enough that it might justify a helper function here:

```go
func gatherFamily(t *testing.T, reg prometheus.Gatherer, familyName string) *dto.MetricFamily {
    t.Helper()
        
    families, err := reg.Gather()
    require.NoError(t, err, "failed to gather metrics")
  
    for _, f := range families {
        if f.GetName() == familyName {
            return f
       }
    }
    
    t.Fatalf("could not find family %s", familyName)
}
```

Reply:
👍